proxy/coco/src/peer.rs

```diff
@@ -123,385 +122,441 @@ impl Peer {
-}
-
-impl Running {
-    /// Abort the tasks of this future.
-    ///
-    /// The next call to [`Future::poll`] will return an [`Error::Aborted`].
-    pub fn abort(&mut self) {
-        self.protocol.1.abort();
-        self.subroutines.1.notify();
-    }
-}
-
-impl From<Peer> for Running {
-    fn from(peer: Peer) -> Self {
-        let Peer {
+        let Self {
             run_loop,
             state,
             store,
             waiting_room,
             subscriber,
             run_config,
-        } = peer;
+        } = self;
 
-        // Note: this must be spawned first to not lose the race for the first
-        // protocol event
+        // Rendezvous on a barrier to let the subroutines subscribe for protocol
+        // events before it actually starts. As `Protocol::subscribe` is async,
+        // we would otherwise need to make `into_running` async as well, which
+        // yields the weird requirement to double `.await` it.
+        let barrier = Arc::new(Barrier::new(2));
         let subroutines = {
-            let stop = Arc::new(tokio::sync::Notify::new());
-            let fut = subroutines(
-                state,
-                store,
-                waiting_room,
-                run_config,
-                subscriber,
-                stop.clone(),
-            );
-            let join = tokio::spawn(fut);
-
-            (join, stop)
-        };
-
-        let protocol = {
-            let (handle, registration) = future::AbortHandle::new_pair();
-            let fut = future::Abortable::new(run_loop, registration);
-            let join = tokio::spawn(fut);
-
-            (join, handle)
+            let barrier = barrier.clone();
+            SpawnAbortable::new(async move {
+                let protocol_events = state.api.protocol().subscribe().await.boxed();
+                barrier.wait().await;
+                Subroutines::new(
+                    state,
+                    store,
+                    waiting_room,
+                    run_config,
+                    protocol_events,
+                    subscriber,
+                )
+                .await
+            })
         };
+        let protocol = SpawnAbortable::new(async move {
+            barrier.wait().await;
+            run_loop.await
+        });
 
-        Self {
+        Running {
             protocol,
             subroutines,
         }
     }
 }
 
+/// Future returned by [`Peer::into_running`].
+#[must_use = "to the sig hup, don't stop, just drop"]
+pub struct Running {
+    /// Join and abort handles for the protocol run loop.
+    protocol: SpawnAbortable<()>,
+    /// The [`Subroutines`] associated with this [`Peer`] instance.
+    subroutines: SpawnAbortable<()>,
+}
+
 impl Drop for Running {
     fn drop(&mut self) {
-        log::debug!("dropping `Peer`");
-        self.abort()
+        log::trace!("`peer::Running` is being dropped");
+        self.protocol.abort()
```

Comment:
Why aren't we aborting the `subroutines`?

Reply:
Oops. Wasn't spawned before 38544772433b69e724c9456ba0405b4328afcc4d, so didn't need to be aborted. Good catch